Core/src/org/sleuthkit/autopsy/geolocation/datamodel/BlackboardArtifactWaypoint.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Autopsy Forensic Browser
+ *
+ * Copyright 2019 Basis Technology Corp.
+ * contact: carrier <at> sleuthkit <dot> org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sleuthkit.autopsy.geolocation.datamodel;
+
+import java.util.Map;
+import org.openide.util.NbBundle.Messages;
+import org.sleuthkit.datamodel.AbstractFile;
+import org.sleuthkit.datamodel.BlackboardArtifact;
+import org.sleuthkit.datamodel.BlackboardAttribute;
+import org.sleuthkit.datamodel.TskCoreException;
+
+/**
+ * Representation of a Waypoint created from a BlackboardArtifact.
+ *
+ */
+public class BlackboardArtifactWaypoint implements Waypoint {
+
+    final private Long timestamp;
+    final private Double longitude;
+    final private Double latitude;
+    final private Double altitude;
+    final private String label;
+    final private AbstractFile image;
+    final private BlackboardArtifact artifact;
+    final private Waypoint.Type type;
+    final private Map<String, String> otherAttributesMap;
+    
+    @Messages({
+        "BlackboardArtifactWaypoint_Last_Known_Label=Last Known Location",
+        "BlackboardArtifactWaypoint_GPS_Trackpoint=GPS Trackpoint",
+        "BlackboardArtifactWaypoint_GPS_Search=GPS Search"
+    })
+
+    /**
+     * Constructs a Waypoint from a BlackboardArtifact
+     *
+     * @param artifact BlackboardArtifact with which to create the waypoint
+     *
+     * @throws TskCoreException
+     */
+    protected BlackboardArtifactWaypoint(BlackboardArtifact artifact) throws TskCoreException {
+        this.artifact = artifact;
+        timestamp = getTimestampFromArtifact(artifact);
+        longitude = GeolocationUtils.getDouble(artifact, BlackboardAttribute.ATTRIBUTE_TYPE.TSK_GEO_LONGITUDE);
+        latitude = GeolocationUtils.getDouble(artifact, BlackboardAttribute.ATTRIBUTE_TYPE.TSK_GEO_LATITUDE);
+        altitude = GeolocationUtils.getDouble(artifact, BlackboardAttribute.ATTRIBUTE_TYPE.TSK_GEO_ALTITUDE);
+        image = getImageFromArtifact(artifact);
+        label = getLabelFromArtifact(artifact);
+        type = getTypeFromArtifact(artifact);
+        otherAttributesMap = GeolocationUtils.getOtherGeolocationAttributes(artifact);
```

Comment:
This is a good solution for the complexity problems pointed out by Codacy, nice!